Fix plot component

ericpre opened the pull request:
This PR resolves #1847.

### Progress of the PR
- [x] Plot convolved component,
- [x] add a note to the user guide to explain the small fitted value when the component are convolved with low loss spectrum,
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix
```python
>>> import hyperspy.api as hs
>>> s = hs.datasets.eelsdb(title="Hexagonal Boron Nitride", spectrum_type="coreloss")[0]
>>> ll = hs.datasets.eelsdb(title="Hexagonal Boron Nitride", spectrum_type="lowloss")[0]
>>> m = s.create_model(ll=ll)
>>> m.enable_fine_structure()
>>> m.smart_fit()
>>> m.plot(plot_components=True)
```

ericpre committed:
Fix bug with plotting individual components for convolved model.

ericpre committed:
Add plotting test for convolved core-loss EELS.

ericpre committed:
Add a note in the documentation to explain the small value of the fitted parameter in case of convolution with the low loss.

ericpre committed:
Tidy up: remove duplicated method.

ericpre committed:
Simplify __call__ method of component.

ericpre committed:
Fix tests component.

thomasaarholt commented:
THIS IS EXCITING

Right. Going to check in the morning.

thomasaarholt commented:
What's up with the tests failing with pngs?

thomasaarholt commented:
This is extremely nice!
Seems to work well, both with and without ll convolution.
<img width="708" alt="screen shot 2018-03-08 at 15 51 54" src="https://user-images.githubusercontent.com/2721423/37157264-a70a3718-22e8-11e8-8b50-e57679007368.png">

ericpre committed:
Use hydrogenic GOS to test model plotting (Hartree-Slater GOS not available in CI).

ericpre committed:
Code tidy up.

thomasaarholt commented on hyperspy/component.py:
```diff
@@ -1027,8 +1020,6 @@ def _component2plot(self, axes_manager, out_of_range2nans=True):
         if not self.active:
             s.fill(np.nan)
-        if self.model.signal.metadata.Signal.binned is True:
-            s *= self.model.signal.axes_manager.signal_axes[0].scale
```
This seems like an important change? Was this relevant just for the convolution plot or plots in general?

ericpre replied:
What do you mean? Regarding binned data, there is no change in this PR.

thomasaarholt replied:
Then why does the PR remove lines 1030 and 1031?

ericpre replied:
Plotting component has been reworked and these lines needed to be removed to be correct. If you read the code, you will understand.

thomasaarholt replied:
Gotcha! Was just making sure you hadn't accidentally removed something :)

thomasaarholt commented:
Thanks for tidying up so much of the code, that sort of spring cleaning is important too!

ericpre committed:
Improve tests coverage.

ericpre committed:
Add files to package_data in setup.py.

francisco-dlp commented on hyperspy/component.py:
```diff
@@ -1013,10 +1009,7 @@ def __call__(self):
-        axis = self.model.axis.axis[self.model.channel_switches]
-        component_array = self.function(axis)
-        return component_array
+        return self.model.__call__(component_list=[self])
```
I think that this is a dangerous change:

*  ``__call__`` should always return the unconvolved component as it was doing previously as the output should not depend on the model and its particular ``convolve`` setting.
* ``__call__`` should work also when ``self.model`` is not defined.

ericpre replied:
Yes, reverted to previous version and move the `model.__call__()` to `_component2plot`.

francisco-dlp commented on hyperspy/models/model1d.py:
```diff
@@ -395,35 +396,33 @@ def __call__(self, non_convolved=False, onlyactive=False):
         """
 
+        if component_list is None:
+            component_list = [component for component in self]
```
``component_list = self``?

francisco-dlp commented on hyperspy/models/model1d.py:
```diff
@@ -395,35 +396,33 @@ def __call__(self, non_convolved=False, onlyactive=False):
 
+        if component_list is None:
+            component_list = [component for component in self]
+        if isinstance(component_list, (list, tuple)):
```
```python
elif not isinstance(component_list, (list, tuple)):
    raise ValueError(
```

francisco-dlp commented on doc/user_guide/eels.rst:
```diff
@@ -149,14 +149,18 @@ Define the chemical composition of the sample
 
-In order to include the effect of plural scattering we provide the low-loss
-spectrum to :py:meth:`~._signals.eels.EELSSpectrum_mixin.create_model`:
+In order to include the effect of plural scattering, the component of the model is convolved with the loss loss spectrum in which case the low loss spectrum needs to be provided to :py:meth:`~._signals.eels.EELSSpectrum_mixin.create_model`:
```
the component of the model is convolved  -> the model is convolved

francisco-dlp commented on doc/user_guide/eels.rst:
```diff
@@ -149,14 +149,18 @@ Define the chemical composition of the sample
 
+.. NOTE::
+
+    In case of the convolution with the low loss spectrum, the fitted value of some of the parameter may seem to be unexpectedly small which can be misleading. Those paramater values are actually correct since the values of the low loss spectrum are usually much larger than 1. Indeed, the `convolution <https://en.wikipedia.org/wiki/Convolution>`_ gives the integral of the pointwise multiplication ...
```
I would remove this note because the reason why the numbers are small is that hyperspy assumes the dwell time of the ll and cl to be the same. We should rather fix that...

ericpre replied:
Removed it. Possibly better to leave this for another PR. Now, we should be parsing the time correctly from DM files, so it should be fine to fix it indeed.

francisco-dlp reviewed (changes requested):
It looks mainly good to me except for the few comments.

ericpre committed:
Simplify model1d.__call__

ericpre committed:
- Fix EELS documentation,
- remove note on convolution with low loss spectrum and corresponding small values of parameters.

ericpre committed:
Revert `component.__call__()` to previous version.

ericpre committed:
Fix test component.

francisco-dlp commented on hyperspy/component.py:
```diff
@@ -1009,15 +1009,20 @@ def __call__(self):
             self.model.axes_manager = axes_manager
             self.fetch_stored_values()
-        s = self.__call__()
+        if self.model:
```
I think that it is unnecessary here to check if model is defined because this method is private and designed to be called only from model. The rest of the code assumes that model is defined.

ericpre replied:
Yes and I think this was already changed in my last commit https://github.com/hyperspy/hyperspy/pull/1851/commits/e884931abfc1a956c789ed534172bf22497fdc75!

francisco-dlp commented:
Except for the minor comment, LGTM!
